File: apps/ios/Pebbles/Features/Record/Steps/RecordNameStep.swift
import SwiftUI

/// Step 2 — what to call it. Clamped to 40 characters as you type, so the
/// counter can never show an over-limit value and there is no error state to
/// design (D3). The limit is front-end only.
struct RecordNameStep: View {
    let name: String
    let limit: Int
    let onChange: (String) -> Void

    /// The counter stays out of the way until the end is in sight: it fades in
    /// at 15 characters left and counts down to 0.
    private static let countdownFrom = 15

    /// The field owns its text so the clamp can write back into it. Feeding
    /// `TextField` a computed `Binding` that truncates in its setter does not
    /// work: UIKit keeps its own buffer during editing, so the extra characters
    /// stay on screen and only vanish at publish — the field says one thing and
    /// the saved pebble says another.
    @State private var text: String
    @FocusState private var isFocused: Bool

    init(name: String, limit: Int, onChange: @escaping (String) -> Void) {
        self.name = name
        self.limit = limit
        self.onChange = onChange
        _text = State(initialValue: name)
    }

    private var remaining: Int { max(0, limit - text.count) }
    private var showsCountdown: Bool { remaining <= Self.countdownFrom }

    var body: some View {
        VStack(spacing: Spacing.sm) {
            TextField("Name", text: $text, axis: .vertical)
                .lineLimit(1...3)
                .pebblesFont(.nameInputHand)
                .multilineTextAlignment(.center)
                .focused($isFocused)
                .submitLabel(.done)
                .padding(Spacing.md)
                .onChange(of: text) { _, newValue in
                    let clamped = String(newValue.prefix(limit))
                    // Assigning state (not a binding setter) is what actually
                    // rewrites the visible field. Guarded so the re-entrant
                    // onChange converges.
                    if clamped != newValue { text = clamped }
                    onChange(clamped)
                }
                .onChange(of: name) { _, newName in
                    // Resync if the model's value moves on its own (a resumed
                    // or reset draft). A no-op while typing, since the model
                    // holds exactly what we just sent it.
                    if newName != text { text = newName }
                }

            // Always laid out, only faded — otherwise the field jumps when the
            // counter appears mid-typing.
            Text(verbatim: "\(remaining)")
                .pebblesFont(.subhead)
                .monospacedDigit()
                .foregroundStyle(
                    remaining == 0 ? Color.accent.primary : Color.system.secondary
                )
                .opacity(showsCountdown ? 1 : 0)
                .animation(.easeInOut(duration: 0.2), value: showsCountdown)
                .accessibilityHidden(!showsCountdown)
                .accessibilityLabel(Text("\(remaining) characters left"))
        }
        .onAppear { isFocused = true }
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/record/steps/RecordNameStep.kt
package app.pbbls.android.features.path.record.steps

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.TextRange
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.text.style.TextAlign
import app.pbbls.android.R
import app.pbbls.android.theme.PebblesText
import app.pbbls.android.theme.PebblesTheme
import app.pbbls.android.theme.PebblesTypography
import app.pbbls.android.theme.Spacing

/**
 * Step 2 — what to call it — ports iOS `RecordNameStep`.
 *
 * A bare field on a bare page: no input background, the handwritten Caveat
 * face, centered, with a countdown that fades in near the limit instead of a
 * permanent counter.
 *
 * **The clamp writes into the state the field itself renders.** Handing
 * `BasicTextField` a [TextFieldValue] whose `onValueChange` truncates *without*
 * writing the truncated value back into local state is the same bug iOS shipped
 * and fixed: the text control keeps its own buffer, so the extra characters stay
 * on screen and are only trimmed at publish — the field says one thing and the
 * saved pebble says another. Local state is the fix; [onChange] carries the
 * already-clamped value up.
 */
@Composable
fun RecordNameStep(
    name: String,
    limit: Int,
    onChange: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    val system = PebblesTheme.colors.system
    val accent = PebblesTheme.colors.accent
    val focusRequester = remember { FocusRequester() }

    var field by remember {
        mutableStateOf(TextFieldValue(text = name, selection = TextRange(name.length)))
    }

    // Resync if the model's value moves on its own (a resumed or reset draft).
    // A no-op while typing, since the model holds exactly what we just sent it.
    LaunchedEffect(name) {
        if (name != field.text) {
            field = field.copy(text = name, selection = TextRange(name.length))
        }
    }

    LaunchedEffect(Unit) { focusRequester.requestFocus() }

    val remaining = (limit - field.text.length).coerceAtLeast(0)
    val showsCountdown = remaining <= COUNTDOWN_FROM
    // Always laid out, only faded — otherwise the field jumps when the counter
    // appears mid-typing.
    val countdownAlpha by animateFloatAsState(
        targetValue = if (showsCountdown) 1f else 0f,
        animationSpec = tween(durationMillis = 200),
        label = "recordNameCountdownAlpha",
    )

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(Spacing.sm),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Box(modifier = Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
            if (field.text.isEmpty()) {
                PebblesText(
                    text = stringResource(R.string.record_name_placeholder),
                    style = PebblesTypography.nameInputHand,
                    color = system.muted,
                    textAlign = TextAlign.Center,
                )
            }
            BasicTextField(
                value = field,
                onValueChange = { incoming ->
                    val clamped = incoming.text.take(limit)
                    // Assigning state (not just reporting upward) is what
                    // actually rewrites the visible field.
                    field =
                        if (clamped == incoming.text) {
                            incoming
                        } else {
                            incoming.copy(
                                text = clamped,
                                selection = TextRange(minOf(incoming.selection.end, clamped.length)),
                            )
                        }
                    onChange(clamped)
                },
                textStyle =
                    PebblesTypography.nameInputHand.copy(
                        color = system.foreground,
                        textAlign = TextAlign.Center,
                    ),
                cursorBrush = SolidColor(accent.primary),
                singleLine = false,
                maxLines = 3,
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Done),
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .padding(Spacing.md)
                        .focusRequester(focusRequester),
            )
        }

        val remainingLabel = stringResource(R.string.record_name_remaining_a11y, remaining)
        PebblesText(
            text = remaining.toString(),
            style = PebblesTypography.subhead,
            color = if (remaining == 0) accent.primary else system.secondary,
            modifier =
                Modifier
                    .alpha(countdownAlpha)
                    .clearAndSetSemantics {
                        if (showsCountdown) contentDescription = remainingLabel
                    },
        )
    }
}

/** The counter stays out of the way until the end is in sight. */
private const val COUNTDOWN_FROM = 15
